elastic/kibana
# [Security Solution][Alerts] Refactor rule execution logic integration tests

## Summary

This PR moves tests that test rule execution logic to a separate folder to test by rule type and refactors most of them (all but 1 for each rule type) to use preview API instead of creating actual rules and waiting for them to run. It does not introduce *new* tests yet, just moving tests to organize them by rule type and improve speed so we can add more rule logic tests easily (and inspect coverage by rule type more easily).

This is possible now that the preview API calls the real rule executor functions.

This PR also fixes a bug with the deduplication logic of threshold rules where it was previously using `signal.rule.rule_id` as the unique identifier for rules. This could cause the preview for threshold rules to incorrectly deduplicate alerts if a specific rule ID parameter is used multiple times for different previews.

Finally, this PR fixes a bug in the `security_solution/alerts/8.0.0` and `security_solution/alerts/8.1.0` archives that prevented them from being loaded.

The slowest remaining part is generally for tests that test exceptions, as we have to create a real exception list which is slow since it's based on saved objects. We could also reduce some esArchive loading/unloading in the future, especially for `auditbeat/hosts` which is shared by many different test files.

The entire `rule_execution_logic` suite now contains 62 tests and takes ~1 minute to execute. Previously, `generating_signals.ts` contained ~35 of these tests and took 5 minutes to execute alone.

To test:
1. In one terminal, start the functional test server: `node scripts/functional_tests_server --config x-pack/test/detection_engine_api_integration/security_and_spaces/rule_execution_logic/config.ts`
2. Once the server is running and available, run the tests in another terminal: `node scripts/functional_test_runner --config x-pack/test/detection_engine_api_integration/security_and_spaces/rule_execution_logic/config.ts`. The `--grep` option can be used to filter for specific tests, e.g. `--grep "EQL type rules"` will only run tests where the description or one of the parent "describe" blocks contains that string. To run a single specific test, include the whole string of `describe` blocks and the `it` block for the test, e.g. `--grep "EQL type rules uses the provided timestamp_field"`

To ensure the refactored tests are not flaky, I used the [flaky test runner](https://ci-stats.kibana.dev/trigger_flaky_test_runner/1) to run the new `rule_execution_logic` suite 100 times (https://buildkite.com/elastic/kibana-flaky-test-suite-runner/builds/1381).
